cli/src/template_parser.rs

```diff
@@ -482,6 +497,28 @@ fn parse_term_node(pair: Pair<Rule>) -> TemplateParseResult<ExpressionNode> {
             })?;
             ExpressionNode::new(ExpressionKind::Integer(value), span)
         }
+        Rule::string_pattern => {
+            let [kind, op, literal] = expr.into_inner().collect_array().unwrap();
+            assert_eq!(kind.as_rule(), Rule::string_pattern_identifier);
+            assert_eq!(op.as_rule(), Rule::pattern_kind_op);
+            let kind = kind.as_str();
+            let text = match literal.as_rule() {
+                Rule::string_literal => STRING_LITERAL_PARSER.parse(literal.into_inner()),
+                Rule::raw_string_literal => parse_raw_string_literal(literal),
+                other => {
+                    return Err(TemplateParseError::expression(
+                        format!("Unexpected literal in string pattern: {other:?}"),
+                        span,
+                    ))
+                }
+            };
+            // XXX(jade): this should probably make the StringPattern
+            // itself here, but because StringPattern is not PartialEq (and
+            // cannot be due to regexes), we defer that to later and Deal With
+            // the AST being slightly less clean (and slightly less validated)
+            // than is ideal.
+            ExpressionNode::new(ExpressionKind::StringPattern { kind, value: text }, span)
```

Comment:
I think the current implementation is correct. We aren't doing any revset/fileset/string pattern parsing in the parser. And we might add other kind of patterns (e.g. date pattern) that share the same syntax.